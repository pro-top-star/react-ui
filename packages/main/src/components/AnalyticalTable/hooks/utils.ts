export const makeTemplateColumns = (columns, resizedColumns) => {
  return columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
};
