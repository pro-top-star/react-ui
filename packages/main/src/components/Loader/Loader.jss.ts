import { JSSTheme } from '../../interfaces/JSSTheme';

//todo use theme colors when available
export const styles = ({ parameters }: JSSTheme) => ({
  '@keyframes scroll': {
    '0%': { backgroundPosition: '-100% 0' },
    '100%': { backgroundPosition: '200% 0' }
  },
  loader: {
    height: '0.25rem',
    width: '100%',
    '&$loaderDeterminate': {
      background: `linear-gradient(to right, ${parameters.sapUiContentIconColor}, ${parameters.sapUiContentIconColor})`,
      backgroundColor: 'rgba(8, 84, 160, 0.15)',
      backgroundRepeat: 'repeat-y'
    },
    '&$loaderIndeterminate': {
      background: `linear-gradient(
      to right,
      rgba(8, 84, 160, 0) 0px,
      rgba(8, 84, 160, 1) calc(50% - 2rem),
      ${parameters.sapUiContentIconColor} calc(50% - 2rem),
      ${parameters.sapUiContentIconColor} calc(50% + 2rem),
      rgba(8, 84, 160, 1) calc(50% + 2rem),
      rgba(8, 84, 160, 0) 100%
    )`,
      backgroundColor: 'rgba(8, 84, 160, 0.15)',
      backgroundSize: '40%',
      backgroundRepeat: 'repeat-y',
      animation: '$scroll 1.2s linear infinite'
    }
  },
  loaderDeterminate: {},
  loaderIndeterminate: {}
});
