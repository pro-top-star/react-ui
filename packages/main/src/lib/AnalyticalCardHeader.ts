import { AnalyticalCardHeader } from '../components/AnalyticalCard/header/AnalyticalCardHeader';

export { AnalyticalCardHeader };
