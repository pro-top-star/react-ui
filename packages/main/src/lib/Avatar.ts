import { Avatar } from '../webComponents/Avatar';

export { Avatar };
