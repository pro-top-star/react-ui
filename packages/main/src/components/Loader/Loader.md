```jsx
import { Loader } from '@ui5/webcomponents-react/lib/Loader';
```
