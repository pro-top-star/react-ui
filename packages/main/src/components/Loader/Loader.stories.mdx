import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs/blocks';
import { Loader } from '@ui5/webcomponents-react/lib/Loader';
import { LoaderType } from '@ui5/webcomponents-react/lib/LoaderType';
import { Card } from '@ui5/webcomponents-react/lib/Card';
import { Text } from '@ui5/webcomponents-react/lib/Text';
import { Icon } from '@ui5/webcomponents-react/lib/Icon';
import { FlexBox } from '@ui5/webcomponents-react/lib/FlexBox';
import { FlexBoxDirection } from '@ui5/webcomponents-react/lib/FlexBoxDirection';
import { createSelectArgTypes } from '@shared/stories/createSelectArgTypes';
import { DocsHeader } from '@shared/stories/DocsHeader';
import { DocsCommonProps } from '@shared/stories/DocsCommonProps';
import { spacing } from '@ui5/webcomponents-react-base';
import '@ui5/webcomponents-icons/dist/activate.js';
import { useEffect, useRef, useState } from 'react';

<Meta
  title="Loading Indicators / Loader"
  component={Loader}
  argTypes={{
    ...createSelectArgTypes({ type: LoaderType }),
    ...DocsCommonProps
  }}
  args={{
    type: LoaderType.Indeterminate,
    progress: '60%',
    style: {},
    className: '',
    tooltip: '',
    slot: '',
    ref: null
  }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      return <Loader {...args} />;
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="." />

<br />

# More Examples

<br />

## Card with Loader

<Canvas>
  <Story name="with Card">
    {(args) => {
      const initialText =
        'This is a text that will be updated 5 seconds after the header is clicked. The text will be visible until the new text has been fetched. To let the user know that the text is updating, the Loader will be displayed.';
      const [text, setText] = useState(initialText);
      const [loading, setLoading] = useState(false);
      const counter = useRef(0);
      const onHeaderClick = () => {
        setLoading(true);
      };
      useEffect(() => {
        if (loading) {
          setTimeout(() => {
            if (counter.current < 10) {
              setText((prev) => (prev === initialText ? 'Updated (⌐■_■)' : initialText));
            } else {
              setText('You really must be bored ಠ_ಠ');
            }
            counter.current++;
            setLoading(false);
          }, 5000);
        }
      }, [loading]);
      return (
        <Card
          avatar={<Icon name="activate" />}
          heading="Click the header to update the text below."
          headerInteractive
          style={{ width: '400px' }}
          onHeaderClick={onHeaderClick}
        >
          <FlexBox direction={FlexBoxDirection.Column}>
            {loading && <Loader {...args} />}
            <Text style={spacing.sapUiContentPadding}>{text}</Text>
          </FlexBox>
        </Card>
      );
    }}
  </Story>
</Canvas>

```jsx
const LoaderComponent = () => {
  const initialText =
    'This is a text that will be updated 5 seconds after the header is clicked. The text will be visible until the new text has been fetched. To let the user know that the text is updating, the Loader will be displayed.';
  const [text, setText] = useState(initialText);
  const [loading, setLoading] = useState(false);
  const onHeaderClick = () => {
    setLoading(true);
  };
  useEffect(() => {
    if (loading) {
      setTimeout(() => {
        setText((prev) => (prev === initialText ? 'Updated (⌐■_■)' : initialText));
        setLoading(false);
      }, 5000);
    }
  }, [loading]);
  return (
    <Card
      avatar={<Icon name="activate" />}
      heading="Click the header to update the text below."
      headerInteractive
      style={{ width: '400px' }}
      onHeaderClick={onHeaderClick}
    >
      <FlexBox direction={FlexBoxDirection.Column}>
        {loading && <Loader />}
        <Text style={spacing.sapUiContentPadding}>{text}</Text>
      </FlexBox>
    </Card>
  );
};
```
