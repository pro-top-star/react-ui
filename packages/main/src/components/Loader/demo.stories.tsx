import { number, select, text } from '@storybook/addon-knobs';
import { Loader } from '@ui5/webcomponents-react/lib/Loader';
import { LoaderType } from '@ui5/webcomponents-react/lib/LoaderType';
import React from 'react';
import notes from './Loader.md';

export const renderLoader = () => (
  <Loader
    type={select('type', LoaderType, LoaderType.Indeterminate)}
    progress={text('progress', '40%')}
    delay={number('delay', 0)}
  />
);
renderLoader.story = {
  name: 'Default'
};

export default {
  title: 'Components / Loader',
  component: Loader,
  parameters: { notes }
};
