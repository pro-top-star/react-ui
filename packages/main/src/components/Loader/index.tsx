import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { LoaderType } from '@ui5/webcomponents-react/lib/LoaderType';
import React, { CSSProperties, FC, forwardRef, RefObject, useEffect, useMemo, useState } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { styles } from './Loader.jss';

export interface LoaderProps extends CommonProps {
  /*
   * Delay in ms until the Loader will be displayed
   */
  delay?: number;
  type?: LoaderType;
  progress?: CSSProperties['width'];
}

const useStyles = createUseStyles<keyof ReturnType<typeof styles>>(styles, { name: 'Loader' });

/**
 * <code>import { Loader } from '@ui5/webcomponents-react/lib/Loader';</code>
 */
const Loader: FC<LoaderProps> = forwardRef((props: LoaderProps, ref: RefObject<HTMLDivElement>) => {
  const { className, type, progress, tooltip, slot, style, delay } = props;
  const classes = useStyles(props);
  const [isVisible, setIsVisible] = useState(delay === 0);

  const loaderClasses = StyleClassHelper.of(classes.loader);
  if (className) {
    loaderClasses.put(className);
  }
  loaderClasses.put(classes[`loader${type}`]);

  const inlineStyles = useMemo(() => {
    const backgroundSize = type !== LoaderType.Determinate ? '40%' : progress;
    return {
      ...style,
      backgroundSize
    };
  }, [progress, style, type]);

  useEffect(() => {
    if (delay > 0) {
      setTimeout(() => {
        setIsVisible(true);
      }, delay);
    }
  }, []);

  const passThroughProps = usePassThroughHtmlProps(props);

  if (!isVisible) {
    return null;
  }

  return (
    <div
      ref={ref}
      className={loaderClasses.valueOf()}
      data-component-name="Loader"
      aria-busy="true"
      role="progressbar"
      title={tooltip || 'Please wait'}
      slot={slot}
      style={inlineStyles}
      {...passThroughProps}
    />
  );
});

Loader.defaultProps = {
  type: LoaderType.Indeterminate,
  progress: '0px',
  delay: 0
};

Loader.displayName = 'Loader';

export { Loader };
