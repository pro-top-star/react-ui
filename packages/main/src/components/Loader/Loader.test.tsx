import { createPassThroughPropsTest, renderThemedComponent } from '@shared/tests/utils';
import { Loader } from '@ui5/webcomponents-react/lib/Loader';
import { LoaderType } from '@ui5/webcomponents-react/lib/LoaderType';
import React from 'react';

describe('Loader', () => {
  test('indeterminate', () => {
    const wrapper = renderThemedComponent(<Loader type={LoaderType.Indeterminate} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('determinate', () => {
    const wrapper = renderThemedComponent(<Loader type={LoaderType.Determinate} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('determinate with progress prop', () => {
    const wrapper = renderThemedComponent(<Loader type={LoaderType.Determinate} progress="42%" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('default', () => {
    const wrapper = renderThemedComponent(<Loader />);
    expect(wrapper).toMatchSnapshot();
  });

  test('with Custom Class Name', () => {
    const wrapper = renderThemedComponent(<Loader className="myTestClass" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('with delay', () => {
    const wrapper = renderThemedComponent(<Loader delay={1000} />);
    expect(wrapper).toMatchSnapshot();
  });

  createPassThroughPropsTest(Loader);
});
