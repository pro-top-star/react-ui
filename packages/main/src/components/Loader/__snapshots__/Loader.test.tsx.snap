// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader default 1`] = `
<div
  aria-busy="true"
  class="Loader--loader- Loader--loaderIndeterminate-"
  data-component-name="Loader"
  role="progressbar"
  style="background-size:40%"
  title="Please wait"
/>
`;

exports[`Loader determinate 1`] = `
<div
  aria-busy="true"
  class="Loader--loader- Loader--loaderDeterminate-"
  data-component-name="Loader"
  role="progressbar"
  style="background-size:0px"
  title="Please wait"
/>
`;

exports[`Loader determinate with progress prop 1`] = `
<div
  aria-busy="true"
  class="Loader--loader- Loader--loaderDeterminate-"
  data-component-name="Loader"
  role="progressbar"
  style="background-size:42%"
  title="Please wait"
/>
`;

exports[`Loader indeterminate 1`] = `
<div
  aria-busy="true"
  class="Loader--loader- Loader--loaderIndeterminate-"
  data-component-name="Loader"
  role="progressbar"
  style="background-size:40%"
  title="Please wait"
/>
`;

exports[`Loader with Custom Class Name 1`] = `
<div
  aria-busy="true"
  class="Loader--loader- myTestClass Loader--loaderIndeterminate-"
  data-component-name="Loader"
  role="progressbar"
  style="background-size:40%"
  title="Please wait"
/>
`;

exports[`Loader with delay 1`] = `null`;
